printer_zpl2/lib/zpl2.py

```diff
@@ -0,0 +1,335 @@
+                'columns_count',
+                'rows_count',
+                'truncate',
+            ]
+            return '7' + self._generateArguments(arguments, kwargs)
+
+        def _ean8(**kwargs):
+            arguments = [
+                'orientation',
+                'height',
+                'interpretation_line',
+                'interpretation_line_above',
+            ]
+            return '8' + self._generateArguments(arguments, kwargs)
+
+        def _upce(**kwargs):
+            arguments = [
+                'orientation',
+                'height',
+                'interpretation_line',
+                'interpretation_line_above',
+                'check_digit',
+            ]
+            return '9' + self._generateArguments(arguments, kwargs)
+
+        def _code128(**kwargs):
+            arguments = [
+                'orientation',
+                'height',
+                'interpretation_line',
+                'interpretation_line_above',
+                'check_digit',
+                'mode',
+            ]
+            return 'C' + self._generateArguments(arguments, kwargs)
+
+        def _ean13(**kwargs):
+            arguments = [
+                'orientation',
+                'height',
+                'interpretation_line',
+                'interpretation_line_above',
+            ]
+            return 'E' + self._generateArguments(arguments, kwargs)
+
+        barcodeTypes = {
+            'code_11': _code11,
+            'interleaved_2_of_5': _interleaved2of5,
+            'code_39': _code39,
+            'code_49': _code49,
+            'pdf417': _pdf417,
+            'ean-8': _ean8,
+            'upc-e': _upce,
+            'code_128': _code128,
+            'ean-13': _ean13,
+        }
+        return '^B' + barcodeTypes[barcodeType](**barcodeFormat)
+
+    def _barcodeFieldDefault(self, barcodeFormat):
+        """ Add the data start command to the buffer """
+        arguments = [
+            'module_width',
+            'bar_width_ratio',
+        ]
+        return '^BY' + self._generateArguments(arguments, barcodeFormat)
+
+    def _fieldDataStart(self):
+        """ Add the data start command to the buffer """
+        return '^FD'
+
+    def _fieldReversePrint(self):
+        """ Allows the printed data to appear white over black, or black over white
+        """
+        return '^FR'
+
+    def _fieldDataStop(self):
+        """ Add the data stop command to the buffer """
+        return '^FS'
+
+    def _fieldData(self, data):
+        """ Add data to the buffer, between start and stop commands """
+        command = ''
+        command += self._fieldDataStart()
+        command += data
+        command += self._fieldDataStop()
+        return command
+
+    def fontData(self, right, down, fieldFormat, data):
+        """ Add a full text in the buffer, with needed formatting commands """
+        command = ''
+        command += self._fieldOrigin(right, down)
```

Comment:
This isn't very efficient. Take a look at this awesome comparison of string concats - https://waymoot.org/home/python_string/

Strings are immutable, so each time you `+=` you're actually creating three strings in memory then discarding two. (old string, add string, concatenated string - discards old and add)